src/home/homework-blob.c

```diff
@@ -0,0 +1,222 @@
+
+        assert(src_fd >= 0);
+        assert(dest_dfd >= 0);
+        assert(name);
+        assert(total_size);
+        assert(uid_is_valid(uid));
+        assert(manifest);
+
+        if (fstat(src_fd, &st) < 0)
+                return log_debug_errno(errno, "Failed to stat fd for %s in blob: %m", name);
+
+        if (S_ISDIR(st.st_mode)) {
+                log_warning("Entry %s in blob directory is a directory. Skipping.", name);
+                return 0;
+        } else if (!S_ISREG(st.st_mode)) {
+                log_warning("Entry %s in blob directory is not a regular file. Skipping.", name);
+                return 0;
+        }
+
+        known_hash = hashmap_get(manifest, name);
+        if (!known_hash) {
+                log_warning("File %s in blob directory is missing from manifest. Skipping.", name);
+                return 0;
+        }
+
+        if (!suitable_blob_filename(name)) {
+                log_warning("File %s in blob directory has invalid filename. Skipping.", name);
+                return 0;
+        }
+
+        initial = lseek(src_fd, 0, SEEK_CUR);
+        if (initial < 0)
+                return log_debug_errno(errno, "Failed to get initial pos on fd for %s in blob: %m", name);
+
+        r = sha256_fd(src_fd, BLOB_DIR_MAX_SIZE, hash);
+        if (r == -EFBIG)
+                return log_warning_errno(SYNTHETIC_ERRNO(EOVERFLOW),
+                                         "Blob directory has exceeded its size limit. Not copying any further.");
+        if (r < 0)
+                return log_debug_errno(r, "Failed to compute sha256 for %s in blob: %m", name);
+
+        size = lseek(src_fd, 0, SEEK_CUR);
+        if (size < 0)
+                return log_debug_errno(errno, "Failed to get final pos on fd for %s in blob: %m", name);
+        if (!DEC_SAFE(&size, initial))
+                return log_debug_errno(SYNTHETIC_ERRNO(EINVAL), "Invalid seek position on fd for %s. Couldn't get size.", name);
+
+        if (lseek(src_fd, initial, SEEK_SET) < 0)
+                return log_debug_errno(errno, "Failed to rewind fd for %s in blob: %m", name);
+
+        if (memcmp(hash, known_hash, SHA256_DIGEST_SIZE) != 0) {
+                log_warning("File %s in blob directory has incorrect hash. Skipping.", name);
+                return 0;
+        }
+
+        if (!INC_SAFE(total_size, size))
+                *total_size = UINT64_MAX;
+        if (*total_size > BLOB_DIR_MAX_SIZE)
+                return log_warning_errno(SYNTHETIC_ERRNO(EOVERFLOW),
+                                         "Blob directory has exceeded its size limit. Not copying any further.");
+
+        WITH_UMASK(0000) {
+                dest = openat(dest_dfd, name, O_WRONLY|O_CREAT|O_EXCL|O_CLOEXEC|O_NOCTTY|O_NOFOLLOW, 0644);
+                if (dest < 0)
+                        return log_debug_errno(errno, "Failed to create/open %s in dest blob: %m", name);
+        }
+
+        r = copy_bytes(src_fd, dest, BLOB_DIR_MAX_SIZE, 0);
```

Comment:
so this of course is racy: the clients could change the file after it was hashed but before we copy it out. In which case we have a corrupted file in place: one that doens't match its supposed hash.

the correct way i guess would be to copy first, and then hash once we are done?

i.e use open_tmpfile_linkable_at() which allows you to use O_TMPFILE for creating the file, and then use link_tmpfile() to actually bind it into the dir if everything checked out.

Reply:
heh that's annoying.

Copy is rather expensive compared to a hash, no? Especially since we're copying across devices, through dm-crypt layers, etc. Perhaps we should hash first, then copy into a tmpfile, then hash a second time to ensure that it didn't change during the copy, then link. That way we skip the expensive copy I/O if the hash never matched in the first place, but still ensure that the client didn't race us